src/dfx/src/lib/migrate.rs

```diff
@@ -0,0 +1,116 @@
+use anyhow::{anyhow, Context, Error};
+use candid::{CandidType, Deserialize, Principal};
+use ic_agent::Identity as _;
+use ic_utils::{
+    call::AsyncCall,
+    interfaces::{
+        management_canister::{
+            builders::{CanisterSettings, InstallMode},
+            StatusCallResult,
+        },
+        ManagementCanister, WalletCanister,
+    },
+    Argument,
+};
+
+use crate::{
+    lib::waiter::waiter_with_timeout,
+    util::{assets::wallet_wasm, expiry_duration},
+};
+
+use super::{
+    environment::Environment,
+    error::DfxResult,
+    identity::{Identity, IdentityManager},
+    models::canister_id_store::CanisterIdStore,
+    network::network_descriptor::NetworkDescriptor,
+    root_key::fetch_root_key_if_needed,
+};
+
+pub async fn migrate(env: &dyn Environment, network: &NetworkDescriptor, fix: bool) -> DfxResult {
+    fetch_root_key_if_needed(env).await?;
+    let agent = env
+        .get_agent()
+        .expect("Could not get agent from environment");
+    let mut mgr = IdentityManager::new(env)?;
+    let ident = mgr.instantiate_selected_identity()?;
+    let wallet = Identity::wallet_canister_id(env, network, ident.name())
+        .map_err(|_| anyhow!("No wallet found; nothing to do"))?;
+    let wallet = WalletCanister::create(agent, wallet)
+        .await
+        .context("No accessible wallet found; nothing to do")?;
+    let mgmt = ManagementCanister::create(agent);
+    if !wallet.version_supports_u128_cycles() {
+        if fix {
+            println!("Upgrading wallet... ");
+            let wasm = wallet_wasm(env.get_logger())?;
+            mgmt.install_code(wallet.canister_id_(), &wasm)
+                .with_mode(InstallMode::Upgrade)
+                .call_and_wait(waiter_with_timeout(expiry_duration()))
+                .await
+                .context("Could not upgrade wallet")?;
+        } else {
+            println!("The wallet is outdated; run `dfx wallet upgrade`");
+        }
+    }
```

Comment:
One of the things that people have been running into is trying to run the 0.9.0 upgrade procedure with an identity that isn't a controller of the wallet.

Can we check the wallet's controllers (available through the equivalent of `dfx canister info`), and if the current identity isn't a controller of the wallet, list identities that are?

Reply:
Generating that list is non-trivial for encrypted identities. It's not difficult to store their principals (and maybe even the principals of their wallets?) in unencrypted format so we have the list available, but it would make it obvious to an attacker which pem file they would want to attempt to crack.

Reply:
Great point!  Similar for HSM.

Reply:
So should this be abandoned?

Reply:
> One of the things that people have been running into is trying to run the 0.9.0 upgrade procedure with an identity that isn't a controller of the wallet.

Can you make it so the tool checks the current identity against the wallet canister's controllers, and displays a nice message about what is wrong and what to do ("This identity isn't a controller of the wallet.  The controllers of the wallet are  [list of controllers].  You need to run this command with an identity that is one of those" or something like that)
 , instead of an error message, which would be more difficult to understand?